src/libcore/slice/mod.rs

```diff
@@ -3130,13 +3151,13 @@ macro_rules! iterator {
                 let mut accum = init;
                 unsafe {
                     while len!(self) >= 4 {
-                        accum = f(accum, & $( $mut_ )* *self.post_inc_start(1))?;
-                        accum = f(accum, & $( $mut_ )* *self.post_inc_start(1))?;
-                        accum = f(accum, & $( $mut_ )* *self.post_inc_start(1))?;
-                        accum = f(accum, & $( $mut_ )* *self.post_inc_start(1))?;
+                        accum = f(accum, self.next_unchecked())?;
+                        accum = f(accum, self.next_unchecked())?;
+                        accum = f(accum, self.next_unchecked())?;
+                        accum = f(accum, self.next_unchecked())?;
```

Comment:
Last time [I touched these](https://github.com/rust-lang/rust/pull/52206), even introducing `inline(always)` functions in some of these places had significant impact on performance. This should get a perf run, and ideally also some manual benchmarking.

Reply:
Yikes, yep, benchmarks I just ran confirm this. So are they not actually being inlined, or what's going on?

Reply:
So we don't need to bother perfbot then?^^

> So are they not actually being inlined, or what's going on?

Your guess is as good as mine. One hypothesis we had last time was that inlining happens after some other passes -- so manually inlined code basically is already present earlier in the pipeline and gets simplified better.

You could try to compile your benchmark to LLVM IR with the old and new version of libstd and compare. You'll see that the function *did* get inlined, but the IR is still vastly different.

Reply:
It's unfortunate that functions apparently aren't zero-cost here, but I guess it is what it is.

I turned `next_unchecked` and `next_back_unchecked` into macros and that seems to have fixed all the regressions. 🙂 It's hard to say for sure though, every time I run the benchmarks there are a few that change by up to 5% or so in either direction, although it's not always the same ones. It seems like they're just not very stable.